apple/Sources/FerrostarCore/ObservableState.swift

```diff
@@ -6,15 +6,15 @@ import UniFFI
 ///
 /// While the core generally does not include UI, this is purely at the model layer and should be implemented
 /// the same for all frontends.
-public final class FerrostarObservableState: ObservableObject {
-    @Published public internal(set) var snappedLocation: CLLocation
-    @Published public internal(set) var heading: CLHeading?
-    @Published public internal(set) var courseOverGround: CLLocationDirection?
-    @Published public internal(set) var fullRouteShape: [CLLocationCoordinate2D]
-    @Published public internal(set) var currentStep: UniFFI.RouteStep?
-    @Published public internal(set) var visualInstructions: UniFFI.VisualInstruction?
-    @Published public internal(set) var spokenInstruction: UniFFI.SpokenInstruction?
-    @Published public internal(set) var distanceToNextManeuver: CLLocationDistance?
+public struct FerrostarObservableState {
```

Comment:
This is a good discussion that I think applies even with the fancier iOS 17+ observable system.

In summary, do we want "state" to be comprised of separate streams that operate independently of each other or do we want to continuously publish a larger struct that represents the whole latest state anytime something changes. 

**@Published single struct**

- State object always contains the latest of everything in a bundle. Easy to consume for nonstandard uses, e.g. as a publisher sink on Car Play or UIKit.
- Less setup required to observe changes (e.g. in custom views).
- I would guess this method makes it easier to export & load as a cached context.

**@Published granular state values**

- The ObservedObject has to be the class on the view `@ObservedObject private var state: FerrostarObservableState` if we want each publisher to trigger updates.
- Each value operates independently, this may be more efficient?, but also makes it tricker to follow along in non standard use cases (opposite as above for CarPlay, UIKit, etc).

**The SwiftUI why**

As for iOS 16 and earlier, because FerrostarCore is the `ObservableObject`, the `@Publisher`s need to be a typed value on that class. If we `ObservableState` to be the class bound to the view that publishes each value as an individual stream, we could do it that way as an alternative. However, a class publishing an `ObservableObject` class with its own publishers won't trigger updates.

Reply:
> In summary, do we want "state" to be comprised of separate streams that operate independently of each other or do we want to continuously publish a larger struct that represents the whole latest state anytime something changes.

Yeah, that's an excellent question. I don't have a strong opinion, but would lean toward "all at once" publishing as it's generally easier, at least until we hit performance issues with SwiftUI. We are currently supporting iOS 16 and earlier, so _the only way we can get more granular UI updates are to break up state_ (or wait till we can support iOS 17+ and the problem goes away with `@Observable`).

As I type this out, I think I realize part of the understanding gap. There are actually _two_ levels that we need to consider: SwiftUI and *waves hand* everything else. You're correctly also thinking about everything else, and I was just thinking about SwiftUI. My above statements about granular updates is correct for SwiftUI but I can see what you mean now for the other contexts. Thanks for pointing that out.

So now, just a few more questions/clarifications, which I think should go quickly:

1. I see that the state _is_ still `@Published` by `FerrostarCore`. I guess this is the core of what drives current observability, right? (It doesn't matter that it's not a ref type with published properties etc.).
2. I don't quite understand how the mutability of the struct fields interacts with publishing. Will changes to properties without assignment to the `observableObject` field on `FerrostarCore` still trigger publishes? For some reason I thought value objects were different but maybe not... This is probably a stupid question, as I've forgotten the finer details of Combine through years of disuse haha. (I would just go out and test this myself by walking around, but I can't in the current state of things :D)
3. Jumping back to `NavigationMapView`, I don't understand why the `navigationState` property is a `var` binding rather than `let`. This is part of what threw me for a loop. I'm fairly certain this can be a `let` now if I understand the machinery, and making it immutable will make everything explicit and ensure we don't shoot ourselves in the foot (mutating this internally would not actually work as expected anyways).
4. Finally, if I have everything straight, we should probably rename the state type to something else. There is nothing inherently observable about the type anymore; it is just state. How does `FerrostarCoreState` sound? Or maybe even something more generic like `NavigationState` since I guess it isn't really specific to anything "core" now that I think about it 😂

Reply:
1. Choosing a struct here over a class is a common practice. It gives you some assurance a modification is going to be published forward. Whereas with a class, there are ways to make a change and not trigger a SwiftUI publish event. This is probably something the new iOS 17+ publishing system just handles. But even with that, I'd say the struct is the right move for the state.
2. Mutability of vars inside the state struct just allows us to do inline modifications without having to create a new init (even though that's basically what swift is doing - copying the struct to a new one with the modification), e.g:

```swift
var newState = self.state
newState.something = newValue
self.state = newState
```

3. It's a requirement of both a `@Published` property. It's likely the new iOS 17+ system massages right over this, but most SwiftUI are typically `var`. See:

<img width="344" alt="Screenshot 2024-01-21 at 1 08 00 PM" src="https://github.com/stadiamaps/ferrostar/assets/4121795/924a3253-5f3d-41a3-89ba-39558fe6064f">

4. Done and I very much like it! I chose the name `state` instead of `navigationState` as seen elsewhere on FerrostarCore because it represents the global state of the navigation system and that keeps it concise when typing. Happy to revise if there's a better reason for `navigationState`.